src/lib/optimizer/strategy/expression_reduction_rule.cpp

```diff
@@ -131,4 +123,49 @@ std::shared_ptr<AbstractExpression> LogicalReductionRule::reduce_distributivity(
   }
 }
 
+std::shared_ptr<AbstractExpression> ExpressionReductionRule::reduce_in_with_single_list_element(
+    const std::shared_ptr<AbstractExpression>& input_expression) {
+  if (const auto in_expression = std::dynamic_pointer_cast<InExpression>(input_expression)) {
+    if (const auto list_expression = std::dynamic_pointer_cast<ListExpression>(in_expression->set())) {
```

Comment:
`const auto&`?

Reply:
We don't do `const auto& a = std::dynamic_pointer_cast` anywhere else. Whatever affection one can find for this reference lifetime extension thingy, I haven't stumbled upon it yet.

Reply:
I just had to check what `dynamic_pointer_cast` returns. Since it returns a new shared_ptr that shares ownership with the original pointer, it doesn't make a difference. If, however, it returned some type of reference, `const auto&` would be the better choice. And that is the beauty of lifetime extension - you can't do anything wrong here. No need to change it, though.